# HG changeset patch
# User StrikerTheHedgefox <mossj32@gmail.com>
# Date 1744307301 10800
#      Thu Apr 10 14:48:21 2025 -0300
# Node ID afab3bf24fb9ad7bddd3356007a95ba3823f1f14
# Parent  fdf62a7ec3e54bada6e4253db377777712e12c4a
# EXP-Topic strikers_stuff
Add +NEVERADJUSTSPRITECLIPPING actor flag.
Prevents the OpenGL renderer from adjusting the sprite out of the floor, regardless of the gl_spriteclip cvar.
Useful for projectiles and particle effects where the sprite adjustment looks terrible, but you still want it for everything else.

diff -r fdf62a7ec3e5 -r afab3bf24fb9 docs/zandronum-history.txt
--- a/docs/zandronum-history.txt	Wed Apr 09 22:36:47 2025 -0300
+++ b/docs/zandronum-history.txt	Thu Apr 10 14:48:21 2025 -0300
@@ -17,6 +17,7 @@
 *+	- Added the ability to apply hardware shaders to cameratextures. Must have a texture in the "textures" folder to replace with the cameratexture for this to work. [StrikerTheHedgefox]
 *+	- Added +DECALUSESTRANSLATION, allowing an actor to pass its translation down to its decal. Useful to change decal color on the fly, or to pass the color of blood particles down to the decal. Pal index 255 is used set the color for shaded decals. [StrikerTheHedgefox]
 *+	- Added +DECALUSESSPRITESCALE, which multiplies an actor's decal scale by the sprite scale. [StrikerTheHedgefox]
+*+	- Added +NEVERADJUSTSPRITECLIPPING, which sets an actor to never adjust the clipping of its sprites to the floor in OpenGL, regardless of the gl_spriteclip setting. [StrikerTheHedgefox]
 *+	- FakeInventory actors can now make use of the "Activation" thing property, vastly increasing its utility. Allows the item to target the player collecting the item and vice/versa, or for the item itself to be the activator. [StrikerTheHedgefox]
 *-	- Send inventory that inherits from CustomInventory and Inventory on full update. Fixes issues where weapon checks for inventory fails for late-joining clients, or after a map change. [StrikerTheHedgefox]
 +	- Hitting ice corpses now trigger damage events. [StrikerTheHedgefox]
diff -r fdf62a7ec3e5 -r afab3bf24fb9 src/actor.h
--- a/src/actor.h	Wed Apr 09 22:36:47 2025 -0300
+++ b/src/actor.h	Thu Apr 10 14:48:21 2025 -0300
@@ -486,6 +486,7 @@
 // [JM] StrikerTheHedgefox's flags
 	HFXF_DECALUSESTRANSLATION		= (1 << 0), // Decal will use actor's translation, or use the translation's pal index 255 color as the basis of its shade if untranslatable.
 	HFXF_DECALUSESSPRITESCALE		= (1 << 1), // Decal's scale will be multiplied by actor's sprite scale.
+	HFXF_NEVERADJUSTSPRITECLIPPING	= (1 << 2), // Never adjust clipping of this actor's sprites in OpenGL, regardless of the gl_spriteclip setting.
 
 // --- mobj.renderflags ---
 
diff -r fdf62a7ec3e5 -r afab3bf24fb9 src/gl/scene/gl_sprite.cpp
--- a/src/gl/scene/gl_sprite.cpp	Wed Apr 09 22:36:47 2025 -0300
+++ b/src/gl/scene/gl_sprite.cpp	Thu Apr 10 14:48:21 2025 -0300
@@ -655,7 +655,7 @@
 		
 		// Tests show that this doesn't look good for many decorations and corpses
 		// [RK] tests also show this looks terrible for floatbob items
-		if (spriteheight>0 && gl_spriteclip>0 && !(thing->flags2 & MF2_FLOATBOB))
+		if (spriteheight>0 && gl_spriteclip>0 && !(thing->flags2 & MF2_FLOATBOB) && !(thing->HFXFlags & HFXF_NEVERADJUSTSPRITECLIPPING))
 		{
 			bool smarterclip = false; // Set to true if one condition triggers the test below
 			if (((thing->player || thing->flags3&MF3_ISMONSTER ||
diff -r fdf62a7ec3e5 -r afab3bf24fb9 src/thingdef/thingdef_data.cpp
--- a/src/thingdef/thingdef_data.cpp	Wed Apr 09 22:36:47 2025 -0300
+++ b/src/thingdef/thingdef_data.cpp	Thu Apr 10 14:48:21 2025 -0300
@@ -278,6 +278,7 @@
 	// [JM] Striker's Flags
 	DEFINE_FLAG(HFXF, DECALUSESTRANSLATION, AActor, HFXFlags),
 	DEFINE_FLAG(HFXF, DECALUSESSPRITESCALE, AActor, HFXFlags),
+	DEFINE_FLAG(HFXF, NEVERADJUSTSPRITECLIPPING, AActor, HFXFlags),
 
 	// [BB] New DECORATE network related flag defines here.
 	DEFINE_FLAG(NETFL, NONETID, AActor, NetworkFlags),
